xformers/components/attention/blocksparse.py

```diff
@@ -133,33 +117,46 @@ def forward(
             q: torch.Tensor,
             k: torch.Tensor,
             v: torch.Tensor,
-            att_mask: Optional[torch.Tensor] = None,
-            key_padding_mask: Optional[torch.Tensor] = None,
             scale: float = 1.0,
             *args,
             **kwargs,
         ) -> torch.Tensor:
+            assert (
+                "att_mask" not in kwargs.keys() and "att_mask" not in args
+            ), "This attention does not support an attention mask, but you can specify causality."
+
             r"""
-            att_mask            A 2D attention mask. The dtype must be the same as q. An additive mask is expected,
-                                meaning float values using "-inf" to mask values.
-            key_padding_mask    A mask with size (batch size x sequence length). The dtype must be the same as q.
-                                An additive mask is expected, meaning float values using "-inf" to mask values
+            A thin wrap around the Triton blockparse attention operation
+
+            .. note: Per element attention mask is not supported, but you can specify causality
             """
 
-            # NOTE:
-            # The attention mask will be taken into account when computing the softmax
-            # meaning that non-masked values which are present in the initial blocksparse layout will be computed.
-            # If blocks are to be constantly masked, better perf would thus be reached by signalling them out in the
-            # initial attention setup
+            # Delayed triton init, to make sure that we get the right device
+            if not hasattr(self, "sparse_dot_sdd"):
```

Comment:
triton blocksparse need the correct device to be passed, but it could be that it's not the case at construction time (if constructed on CPU then moved), and it's not possible to just default to cuda:0 (would break many multi-gpu cases). So we defer the construction until after the first input tensor comes in

Reply:
sounds good to me.

One other option would be to inherit the `.cuda()` / `.to()` methods so that they re-create those objects.

The current approach is fine as is because the those objects don't contain learnable parameters, but if that was the case it would mess up with optimizers / distributed.

Reply:
yep, that's something we discussed with @colehawkins on another PR, one possible issue for me is that some sharded trainers intercept the .to() calls, so this would silently fail in that case. I think that both takes have issues (the delayed init and the .to() overload), the only clean way out that I can think of is to make this attention take a "device" as a construction argument, and put it to the right place from the beginning

Reply:
@blefaudeux @fmassa  I've been using this with both pytorch lightning and the huggingface trainer, and this method (initialization at first forward) is the cleanest way I found that doesn't break any standard workflows or require model initialization workarounds. If we just take the device at construction this breaks the huggingface trainer "natural approach" for single-node, multi-gpu which (1) create the model, then (2) call `.to()`.

One alternative is to initialize with a device at construction, keep that as `self.device`, and then check that against `query.device` and possibly re-initialize at the forward pass.